src/unix/udp.c

```diff
@@ -654,28 +654,71 @@ int uv__udp_connect(uv_udp_t* handle,
   return 0;
 }
 
-
+/* From https://pubs.opengroup.org/onlinepubs/9699919799/functions/connect.html
+ * Any of uv supported UNIXs kernel should be standardized, but the kernel 
+ * implementation logic not same, let's use pseudocode to explain the udp
+ * disconnect behaviors:
+ * 
+ * Predefined stubs for pseudocode:
+ *   1. sodisconnect: The function to perform the real udp disconnect
+ *   2. pru_connect: The function to perform the real udp connect
+ *   3. so: The kernel object match with socket fd
+ *   4. addr: The sockaddr parameter from user space
+ * 
+ * BSDs:
+ *   if(sodisconnect(so) == 0) { // udp disconnect succeed
+ *     if (addr->sa_len != so->addr->sa_len) return EINVAL;
+ *     if (addr->sa_family != so->addr->sa_family) return EAFNOSUPPORT;
+ *     pru_connect(so);
+ *   }
+ *   else return EISCONN;
+ *
+ * z/OS (same with Windows):
+ *   if(addr->sa_len < so->addr->sa_len) return EINVAL;
+ *   if (addr->sa_family == AF_UNSPEC) sodisconnect(so);
+ *
+ * AIX:
+ *   if(addr->sa_len != sizeof(struct sockaddr)) return EINVAL; // ignore ip proto version
+ *   if (addr->sa_family == AF_UNSPEC) sodisconnect(so);
+ *
+ * Linux,Others:
+ *   if(addr->sa_len < sizeof(struct sockaddr)) return EINVAL;
+ *   if (addr->sa_family == AF_UNSPEC) sodisconnect(so);
+ */
 int uv__udp_disconnect(uv_udp_t* handle) {
     int r;
+#if defined(__MVS__)
+    struct sockaddr_storage addr;
+#else
     struct sockaddr addr;
+#endif
 
     memset(&addr, 0, sizeof(addr));
-
+    
+#if defined(__MVS__)
+    addr.ss_family = AF_UNSPEC;
+#else
     addr.sa_family = AF_UNSPEC;
-
+#endif
+    
     do {
       errno = 0;
-      r = connect(handle->io_watcher.fd, &addr, sizeof(addr));
+      r = connect(handle->io_watcher.fd, (struct sockaddr*) &addr, sizeof(addr));
     } while (r == -1 && errno == EINTR);
 
-    if (r == -1 && errno != EAFNOSUPPORT)
+    if (r == -1) {
+#if defined(BSD) // The macro BSD is from sys/param.h
+      if (errno != EAFNOSUPPORT && errno != EINVAL)
```

Comment:
I'm not sure it's okay to ignore EINVAL, that means the socket hasn't been actually disconnected, right?

I would expect the logic to look something like this, i.e., try different address families:
```c
static const socklen_t sizes[] = {
  sizeof(struct sockaddr_in),
  sizeof(struct sockaddr_in6),
  sizeof(struct sockaddr_un)
};

struct sockaddr_storage s;
struct sockaddr* p;
socklen_t* n;

memset(&s, 0, sizeof(s));
s.ss_family = AF_UNSPEC;
p = (struct sockaddr*) &s;

for (n = sizes; n < sizes + ARRAY_SIZE(sizes); n += 1) {
  do
    r = connect(fd, p, *n);
  while (r == -1 && errno == EINTR);

  if (r == 0)
      return 0;

  if (errno != EAFNOSUPPORT)
    if (errno != EINVAL)
      return -1;
}

return -1;
```

Reply:
It's not right, the EINVAL not means the socket hasn't been actually disconnected. Both `EINVAL` and `EAFNOSUPPORT` caused by the BSD kernel  always try to do socket connect after do disconnected succeed.